code/game/objects/structures/doors.dm

```diff
@@ -0,0 +1,902 @@
+		deconstruct(TRUE)
+
+///Build the door
+/obj/structure/door/proc/do_construction()
+	var/number_of_steps = length(construction_steps)
+	construction_steps_index--
+
+	if(construction_steps_index != number_of_steps - 1)
+		name = base_name
+		anchored = initial(anchored)
+
+/obj/structure/door/proc/toggle_lock(mob/user, obj/item/key/door/key)
+	if(!lock)
+		balloon_alert(user, "No lock installed")
+		return
+
+	lock.toggle(user, key)
+
+/obj/structure/door/Cross(atom/movable/AM)
+	. = ..()
+	if(.)	//Non-full tile doors will not block the crosser, no need to continue below
+		if(CHECK_BITFIELD(atom_flags, ON_BORDER) && REVERSE_DIR(dir) == AM.dir)	//Auto close non-full tile doors if the crosser walked through the door
+			auto_close(AM)
+		return .
+
+	//Don't care about the user if they're not a mob type
+	attempt_to_open(ismob(AM) ? AM : null, direction_from_opener = angle2dir(Get_Angle(src, AM)), bumped = TRUE)
+	var/is_open = CHECK_BITFIELD(door_flags, DOOR_OPEN)
+	if(is_open && CHECK_BITFIELD(atom_flags, ON_BORDER) && REVERSE_DIR(dir) == AM.dir)
+		auto_close(AM)
+
+	return is_open
+
+///Mob opening the door has no left leg
+#define OPENER_MISSING_LEFT_LEG 1<<0
+///Mob opening the door has no right leg
+#define OPENER_MISSING_RIGHT_LEG 1<<1
+///Mob opening the door has no unbroken legs
+#define OPENER_ALL_LEGS_CRIPPLED 1<<2
+
+///To prevent doing 5 kicks a second; will return FALSE if the cooldown is still active, TRUE if it is not and will start the cooldown
+/obj/structure/door/proc/handle_door_force_open_cooldown(mob/door_opener, cooldown = 1.5 SECONDS)
+	if(TIMER_COOLDOWN_CHECK(door_opener, COOLDOWN_FORCE_OPEN_DOOR))
+		return FALSE
+
+	TIMER_COOLDOWN_START(door_opener, COOLDOWN_FORCE_OPEN_DOOR, cooldown)
+	return TRUE
+
+///To prevent near instantly killing people with door slams; will return FALSE if the cooldown is still active, TRUE if it is not and will start the cooldown
+/obj/structure/door/proc/handle_door_force_close_cooldown(mob/door_opener, cooldown = 1 SECONDS)
+	if(TIMER_COOLDOWN_CHECK(door_opener, COOLDOWN_FORCE_CLOSE_DOOR))
+		return FALSE
+
+	TIMER_COOLDOWN_START(door_opener, COOLDOWN_FORCE_CLOSE_DOOR, cooldown)
+	return TRUE
+
+///Separated this proc so that individual doors can override this if they want to change how it is forced open; do not call this, use attempt_to_open()
+/obj/structure/door/proc/force_door_open(mob/user, forced, damage = 10, bumped, leg_flags)
+	if(!forced && user && !handle_door_force_open_cooldown(user))
+		return FALSE
+
+	playsound(src, slam_sound, 70, FALSE, 10, 2)
+	if(!CHECK_BITFIELD(door_flags, DOOR_FORCEABLE))
+		if(user)
+			to_chat(user, span_warning("[src] is too sturdy for brute force!"))
+		return FALSE
+
+	//Do all of this if there is a door kicker/tackler
+	if(user)
+		//If not tackling the door, that means they are attempting to kick it with a broken leg, so let's see if we got any legs to break for being a fool
+		if(!bumped && CHECK_BITFIELD(leg_flags, OPENER_ALL_LEGS_CRIPPLED))
```

Comment:
early return?

Reply:
I didn't make it early return since both chains eventually have to do a `return TRUE` by the end.